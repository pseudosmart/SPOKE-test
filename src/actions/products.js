import { REQUEST_PRODUCTS, RECEIVE_PRODUCTS } from '../constants/actions'
import { graphQL } from '../helpers'

export const fetchProducts = () => dispatch => {
  dispatch({type: REQUEST_PRODUCTS})
  return graphQL.query().then(products => {
    return dispatch(productFetchSuccess(products))
  })
}

export const productFetchSuccess = products => ({
  type: RECEIVE_PRODUCTS,
  payload: products
})

export const productFetchError = error => ({ //TODO
  type: RECEIVE_PRODUCTS,
  error: true,
  payload: error
})
