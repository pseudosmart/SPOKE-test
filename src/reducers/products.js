import { REQUEST_PRODUCTS, RECEIVE_PRODUCTS } from '../constants/actions'

const flattenProducts = products => {
  const flatProducts = products.data.products.edges.reduce((productList, product) => { // Would need much better error handling
    const {id, images, productType, title, tags, priceRange, updatedAt} = product.node
    return [
      ...productList,
      {
        id,
        image: images.edges.length ? images.edges.pop().node.originalSrc : '',
        productType,
        title,
        tags,
        price: parseInt(priceRange.maxVariantPrice.amount, 10).toFixed(2),
        updatedAt
      }
    ]
  }, [])

  return flatProducts
}

export default (currentState = {}, action) => {
  switch (action.type) {
    case REQUEST_PRODUCTS:
      return {...currentState, fetching: true}

    case RECEIVE_PRODUCTS:
      const formattedProducts = flattenProducts(action.payload)
      return { ...currentState, products: formattedProducts, fetching: false} //TODO use the fetching
      
      
      
    default:
      return currentState
  }
}