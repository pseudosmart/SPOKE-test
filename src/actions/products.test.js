import * as actions from './products'

let dispatch

beforeEach(() => {
  dispatch = jest.fn()
})

test('fetchProducts', () => {
  expect(actions.fetchProducts()(dispatch)).toMatchSnapshot()
  expect(dispatch).toHaveBeenCalled()
})

test('productFetchSuccess', () => {
  expect(actions.productFetchSuccess([{ id: 'test', title: 'title' }, { id: 'test2', title: 'title' }])).toMatchSnapshot()
})

test('productFetchError', () => {
  expect(actions.productFetchSuccess('Error')).toMatchSnapshot()
})
