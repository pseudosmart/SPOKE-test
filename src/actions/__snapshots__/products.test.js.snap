// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchProducts 1`] = `Promise {}`;

exports[`productFetchError 1`] = `
Object {
  "payload": "Error",
  "type": "RECEIVE_PRODUCTS",
}
`;

exports[`productFetchSuccess 1`] = `
Object {
  "payload": Array [
    Object {
      "id": "test",
      "title": "title",
    },
    Object {
      "id": "test2",
      "title": "title",
    },
  ],
  "type": "RECEIVE_PRODUCTS",
}
`;
